Add references to Nbshell in export templates, reports and scripts

8ctorres opened the pull request:
### Fixes: #15014

The Netbox shell is a very powerful resource for people trying to develop their own scripts, reports or configuration templates, but I feel it is difficult to find in the documentation. For this reason, I have added a few references to it in the pages, so that a new user can find it much sooner than it took me to find.

8ctorres committed:
Add references to Nbshell in export templates, reports and scripts

8ctorres committed:
Add references to Nbshell in export templates, reports and scripts

jeremystretch commented on docs/administration/netbox-shell.md:
```diff
@@ -1,6 +1,6 @@
 # The NetBox Python Shell
 
-NetBox includes a Python management shell within which objects can be directly queried, created, modified, and deleted. To enter the shell, run the following command:
+NetBox includes a Python management shell within which objects can be directly queried, created, modified, and deleted. To enter the shell, run the following command from a shell that has the netbox virtualenv activated:
```
```suggestion
NetBox includes a Python management shell within which objects can be directly queried, created, modified, and deleted. To enter the shell, run the following command with the appropriate Python virtual environment activate:
```

jeremystretch commented on docs/administration/netbox-shell.md:
````diff
@@ -143,6 +143,40 @@ To return the inverse of a filtered queryset, use `exclude()` instead of `filter
 346
 ```
 
+If the query returns only one object, the get() method can be used. This method will yield the actual object resulting from the query, instead of a QuerySet. For this to work, the query must return only one object. The syntax is identical to the filter and exclude methods. For example, we can get a device from it's asset tag:
+
+```
+>>>
+>>> Device.objects.get(asset_tag="100079912515")
+<Device: AP994003 (100079912515)>
+>>>
+```
+
+If the query returns more than one object, a MultipleObjectsReturned exception will be thrown:
+
+```
+>>> Device.objects.get(role_id=13)
+Traceback (most recent call last):
+  File "<console>", line 1, in <module>
+  File "/srv/netbox/venv/lib/python3.10/site-packages/django/db/models/manager.py", line 87, in manager_method
+    return getattr(self.get_queryset(), name)(*args, **kwargs)
+  File "/srv/netbox/venv/lib/python3.10/site-packages/django/db/models/query.py", line 640, in get
+    raise self.model.MultipleObjectsReturned(
+dcim.models.devices.Device.MultipleObjectsReturned: get() returned more than one Device -- it returned more than 20!
+>>>
+```
+
+Queries can all also be executed from a particular object instead of from the model itself. For instance, to get all circuits that are assigned to one site, it is easier to filter from the site itself, instead of using the "Circuit" model and building the query from there. This is particularly useful for configuration templates and export templates, since it allows to query other database objects ...
+
+```
+>>> site.circuit_terminations.all()
+<RestrictedQuerySet [<CircuitTermination: 20899518: Termination A>, <CircuitTermination: DT00018356: Termination A>]>
+>>>
+```
+
+The same methods (all, filter, exclude, get...) can be used in this kind of queries.
+
+
````
This is out of scope for the NetBox documentation as it describes generic behavior of the [Django ORM](https://docs.djangoproject.com/en/stable/topics/db/queries/). Rather than duplicate Django documentation, we should refer users there to read further.

8ctorres replied:
Done. I reduced it to just one line about the get() method, and left the tip admonition pointing users to the Django documentation page

jeremystretch commented on docs/customization/custom-scripts.md:
````diff
@@ -56,6 +56,15 @@ class AnotherCustomScript(Script):
 script_order = (MyCustomScript, AnotherCustomScript)
 ```
 
+### The run() method
````
Always use markup to highlight code excerpts, and capitalize heading titles.
```suggestion
### The `run()` Method
```

jeremystretch commented on docs/customization/custom-scripts.md:
```diff
@@ -56,6 +56,15 @@ class AnotherCustomScript(Script):
 
+### The run() method
+
+The run() method is the entrypoint for the script, and it runs in the context of Netbox's own execution environment. This means from here, everything inside Netbox itself is accesible. The [Netbox Shell](../administration/netbox-shell.md) is a good resource to keep in hand, since it allows to see the objects in Netbox in the same way the run() method of the script does.
```
NetBox is always written as either "NetBox" (proper usage) or "netbox" (referencing code/resources); never "Netbox."

8ctorres replied:
Done. Sorry for the mistake

jeremystretch commented on docs/customization/export-templates.md:
````diff
@@ -32,6 +32,8 @@ If you need to use the config context data in an export template, you'll should
 {% endfor %}
 ```
 
+To see all the attributes of a given object, you can use the [Netbox Shell](../administration/netbox-shell.md). It supports autocompletion and allows one to see all of the methods and attributes a given object type has. All of them can be called from within a Jinja template. Using queries from one object to another, one can navigate pretty much the entire Netbox object model. For instance, from a ...
````
This should be reduced to a "tip" admonition as the topic strays from export templates. Just point out that object attributes can be explored using `nbshell`.

jeremystretch commented on docs/customization/reports.md:
````diff
@@ -29,7 +29,9 @@ class DeviceIPsReport(Report):
 ```
 
-Within each report class, we'll create a number of test methods to execute our report's logic. In DeviceConnectionsReport, for instance, we want to ensure that every live device has a console connection, an out-of-band management connection, and two power connections.
+Within each report class, we'll create a number of test methods to execute our report's logic. The method's name must start with "test_" and it takes no arguments.
````
```suggestion
Within each report class, we'll create a number of test methods to execute our report's logic. The name of each method must begin with `test_` and the method must accept no arguments.
```

jeremystretch commented on docs/customization/reports.md:
```diff
@@ -82,6 +84,8 @@ class DeviceConnectionsReport(Report):
 
 As you can see, reports are completely customizable. Validation logic can be as simple or as complex as needed. Also note that the `description` attribute support markdown syntax. It will be rendered in the report list page.
 
+In the same way scripts do, reports run from within Netbox's own environment and can access the objects inside Netbox directly. The [Netbox Shell](../administration/netbox-shell.md) is a good resource to keep in hand, since it allows to see the objects in Netbox in the same way the test methods of a report do.
```
Please correct "Netbox" to NetBox. This should also probably render as a tip admonition.

8ctorres replied:
Done

jeremystretch commented on docs/features/configuration-rendering.md:
````diff
@@ -35,6 +35,8 @@ Configuration templates are written in the [Jinja2 templating language](https://
 {% endblock %}
 ```
 
+To see all the attributes of a given object, you can use the [Netbox Shell](../administration/netbox-shell.md). It supports autocompletion and allows one to see all of the methods and attributes a given object type has. All of them can be called from within the configuration template. Also, other objects, for instance, a device's interfaces or connected circuits can be accessed from the template  ...
````
As with export templates, lets compress this into a brief tip admonition pointing the reader to the `nbshell` page.

8ctorres replied:
Done

jeremystretch reviewed (changes requested):
Thanks @8ctorres. Let's compress some of the commentary on using `nbshell` to ensure we don't disrupt from the primary flow of each page.

8ctorres commented:
Hi @jeremystretch thanks for your feedback. I will try to make some changes this weekend and update the PR.

jeremystretch commented:
@8ctorres checking in. Are you still working on this PR?

8ctorres commented:
Hi @jeremystretch thanks for checking. I haven't really had the time these past few weeks, I apologize for the delay. I will try to get this done before the end of the month. Sorry again

8ctorres committed:
Corrections in documentation about nbshell

8ctorres committed:
Corrections in documentation about nbshell

8ctorres commented:
Hi @jeremystretch , sorry again for the delay with this. I have reviewed your suggestions and added another commit with a few changes. About reports, I see that they're being deprecated in v4, so maybe the documentation on that section isn't going to be so relevant anymore.

Thanks!

arthanson commented:
@8ctorres can you please fix the merge conflicts

8ctorres committed:
Merge branch 'develop' into develop

8ctorres commented:
Hi @arthanson sorry for the delay. I fixed the merge conflicts in the custom scripts page. For the custom reports page, since they're being deprecated anyways, I just rescinded all my changes and kept the incoming ones. I hope that's fine

arthanson committed:
16586 add .python-version to gitignore

github-actions[bot] committed:
Update source translation strings

github-actions[bot] committed:
Update source translation strings

jeremystretch committed:
Fixes #16542: Fix bulk form operations when HTMX is enabled

jeremystretch committed:
Fixes #16452: Fix sizing of buttons within object attribute panels

jeremystretch committed:
Fixes #16450: Rack unit filter should be case-insensitive

arthanson committed:
16460 remove spaces from telephone dialing

jeremystretch committed:
Fixes #16444: Disable ordering circuits list by A/Z termination

github-actions[bot] committed:
Update source translation strings

jeremystretch committed:
Fixes #16512: Restore a user's preferred language on login (#16628)

arthanson committed:
16416 enable dark/light toggle in mobile view (#16635)

* 16416 enable dark/light toggle in mobile view

* 16416 move to inc file

arthanson commented:
@8ctorres Looks good, can you remove the changes in netbox/translations/en/LC_MESSAGES/django.po not sure why or how they got there

Julio-Oliveira-Encora committed:
Feature 15348 - Quick Access Saved Filters (#15862)

* Added dropdown for Saved Filters.

* Added dropdown for Saved Filters.

* Added dropdown for Saved Filters.

* Fixed linter issues in savedFiltersSelect.ts

* Fixed linter issues in netbox.ts

* Fixed linter issues in netbox.ts

* Removed the blue tag with the filters when saved filters is selected.

* Adjusts in table_controls_htmx.html to vertical height of the Quick Search match to the dropdown.

* Adjusts in table_controls_htmx.html to vertical height of the Quick Search match to the dropdown.

* Adjusts in table_controls_htmx.html to vertical height of the Quick Search match to the dropdown.

* Minor adjusts in savedFiltersSelect.ts

* Addressed PR comment.

* Addressed PR comment.

* Addressed PR comment.

* Omit saved filters from 'applied filters'; clean up form widget

---------

Co-authored-by: Jeremy Stretch <jstretch@netboxlabs.com>

jeremystretch committed:
Update changelog for #15348, #16416, #16444, #16450, #16452, #16460, #16512, #16542

github-actions[bot] committed:
Update source translation strings

8ctorres committed:
Revert "Update source translation strings"

This reverts commit 6da24c909fd945e7efa260ab1b535d58ac53c1f5 which was auto generated by github-bot

8ctorres commented:
Hi @arthanson, I think it must be some automatic github action that updates translation/locale metadata. In any case, I just reverted that commit so it should be good now.

github-actions[bot] committed:
Update source translation strings

jeffgdotorg committed:
Interim fix to SECURITY.md: Remove non-working email address

jeremystretch commented on docs/administration/netbox-shell.md:
````diff
@@ -105,6 +105,8 @@ This approach can span multiple levels of relations. For example, the following
 >>> IPAddress.objects.filter(interface__device__site__region__slug="north-america")
 ```
 
+The `get()` method can be used in place of the `filter()` method if we expect this query to return only one object. This method will yield the actual object resulting from the query, instead of a QuerySet. If the query returns more than one result, it will raise an exception.
````
This is redundant to information provided in the [Django documentation](https://docs.djangoproject.com/en/5.0/ref/models/querysets/). Please reference it here instead of duplicating text that we'll need to then maintain.

jeremystretch commented on docs/administration/netbox-shell.md:
````diff
@@ -143,6 +145,8 @@ To return the inverse of a filtered queryset, use `exclude()` instead of `filter
 346
 ```
 
+Queries can all also be executed from a particular object instead of from the model itself. For instance, to get all circuits that are assigned to one site, it is easier to filter from the site itself, instead of using the "Circuit" model and building the query from there. This is particularly useful for configuration templates and export templates, since it allows to query other database objects ...
````
Again, this is redundant to the Django documentation, which provides plenty of context and examples.

jeremystretch commented on docs/customization/custom-scripts.md:
````diff
@@ -65,6 +65,18 @@ class AnotherCustomScript(Script):
 script_order = (MyCustomScript, AnotherCustomScript)
 ```
 
+The `run()` method is the entrypoint for the script, and it runs in the context of NetBox's own execution environment. This means from here, everything inside NetBox itself is accesible. The [NetBox Shell](../administration/netbox-shell.md) is a good resource to keep in hand, since it allows to see the objects in NetBox in the same way the script does.
````
We've already introduced the `run()` method above; this makes no sense to put here. The link to the `nbshell` documentation should be a tip admonition as it doesn't directly relate to authoring a script.

jeremystretch commented on docs/customization/custom-scripts.md:
```diff
@@ -65,6 +65,18 @@ class AnotherCustomScript(Script):
+The `run()` method can itself call other methods that are in the same module but outside the "MyCustomScript" class, and if there are several scripts in the same module (this is, in the same Python file), both scripts can reuse the same auxiliary methods, keeping the code cleaner. For this reason, it is encouraged to keep similar scripts in the same module. Script can return a string, which will  ...
+
+
+## Module Attributes
+
+### `name`
+
+You can define `name` within a script module (the Python file which contains one or more scripts) to set the module name. If `name` is not defined, the module's file name will be used.
+
+
```
This is no longer supported.
```suggestion
```

jeremystretch reviewed (changes requested):
@8ctorres could you please rebase your changes against the current `develop` branch and remove the unrelated commits? (I tired to myself but you have not given access for PR reviewers to push to your branch.)

nopg committed:
Fixes #15717 - Unable to assign a VM in Site to Cluster without Site (#15763)

* Fixes #15717: Allow VM with Site to Cluster without Site

* Fixes #15717: Allow VM with Site to Cluster without Site

* Fixes #15717: Allow VM with Site to Cluster without Site

* Fixes #15717: Allow VM with Site to Cluster without Site

* Fixes #15717: Allow VM with Site to Cluster without Site

tobiasge committed:
Closes #16273: Add search box to menu on mobile

jeffgdotorg committed:
Remove dead link to project-stats anchor. Fixes #16621

github-actions[bot] committed:
Update source translation strings

arthanson committed:
16252 only show results count if paginator (#16269)

* 16252 only show results count if paginator

* 16252 hack in table page count

* 16252 review changes

arthanson committed:
16149 add (optional) obj hyperlink to script list table (#16271)

* 16149 add (optional) obj hyperlink to script list table

* 16149 add (optional) obj hyperlink to script list table

* 16149 review feedback

* 16149 review changes

siku4 committed:
fix: add missing parent field to inventory item import form

github-actions[bot] committed:
Update source translation strings

jeremystretch committed:
Closes #16686: Relete obsolete OpenAPI definitions

jeremystretch committed:
Closes #16367: Update census URL

jeremystretch committed:
Closes #16307: Enable calling log_* methods on Script without a log message

jeremystretch committed:
Closes #16700: Audit usage of mark_safe() for consistent escaping

jeremystretch committed:
Fixes #16702: Fix validation of return_url query parameter

jeremystretch committed:
Changelog for #15717, #16149, #16252, #16273, #16307, #16702

transifex-integration[bot] committed:
Updates for project NetBox (#16687)

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in pt

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in fr [Manual Sync]

12% of minimum 1% reviewed source file: 'django.po'
on 'fr'.

Sync of partially translated files: 
untranslated content is included with an empty translation 
or source language content depending on file format

* Translate django.po in pt [Manual Sync]

100% reviewed source file: 'django.po'
on 'pt'.

* Translate django.po in ru [Manual Sync]

30% of minimum 1% reviewed source file: 'django.po'
on 'ru'.

Sync of partially translated files: 
untranslated content is included with an empty translation 
or source language content depending on file format

* Translate django.po in de [Manual Sync]

75% of minimum 1% reviewed source file: 'django.po'
on 'de'.

Sync of partially translated files: 
untranslated content is included with an empty translation 
or source language content depending on file format

* Translate django.po in tr [Manual Sync]

7% of minimum 1% reviewed source file: 'django.po'
on 'tr'.

Sync of partially translated files: 
untranslated content is included with an empty translation 
or source language content depending on file format

---------

Co-authored-by: transifex-integration[bot] <43880903+transifex-integration[bot]@users.noreply.github.com>

jeremystretch committed:
Release v4.0.6

jeremystretch committed:
Recompile JS assets

jeremystretch committed:
Merge pull request #16707 from netbox-community/develop

Release v4.0.6

jeremystretch committed:
PRVB

github-actions[bot] committed:
Update source translation strings

jeffgdotorg committed:
Added missing CDN cache clearing step to release checklist in docs

Julio-Oliveira-Encora committed:
16704 - Define a default help_text for ColorField (#16708)

* Added `help_text` to ColorField.

* Addressed PR comment to remove the redundant help_text from all the forms where ColorField was used.

* Add space before example value

---------

Co-authored-by: Jeremy Stretch <jstretch@netboxlabs.com>

Julio-Oliveira-Encora committed:
Was added to searching support languages other than English for objec… (#16706)

* Was added to searching support languages other than English for object types(s).

* Fix SearchForm field label translation

---------

Co-authored-by: Jeremy Stretch <jstretch@netboxlabs.com>

Julio-Oliveira-Encora committed:
16424 - Allow filtering of Devices by Cluster and Cluster Group (#16674)

* Allow filtering Devices by Cluster and Cluster Group.

* Allow filtering Devices by Cluster and Cluster Group.

* Added tests for cluster and cluster_groups filterset.

* Add missing filter & complete tests

---------

Co-authored-by: Jeremy Stretch <jstretch@netboxlabs.com>

prryplatypus committed:
Quote `VIRTUALENV`

github-actions[bot] committed:
Update source translation strings

tobiasge committed:
Fixes #16689: Load correct configuration

Loads the the current configuration if no ConfigRevisions are saved to
the database.

Julio-Oliveira-Encora committed:
Added default:"0" to total_count in object_list.html

peteeckel committed:
Updated the documentation section about removing plugins

Julio-Oliveira-Encora committed:
16725 - The admin section should always come last in the navigation menu (#16762)

* I replaced `append` with `insert` into menu.py to make the admin section appear last in the navigation menu.

* Clean up ordering logic

---------

Co-authored-by: Jeremy Stretch <jstretch@netboxlabs.com>

jeremystretch committed:
Fixes #16523: Restore highlighting of current device in virtual chassis members panel

github-actions[bot] committed:
Update source translation strings

peteeckel committed:
Do not delete all search indexes when reindexing specific models (#16755)

* Do not delete all search indexes when reindexing specific models

* Clear all indexes only if neither --lazy nor a list of models are
  specified for "manage.py reindex"

* Otherwise, clear the index for a model immediately before rebuilding
  it

* Separated clearing from re-indexing the search cache

llamafilm committed:
fix: allow cloning field value of 0 (#16741)

* fix: allow cloning field value of 0

* Fix evaluation of False value

---------

Co-authored-by: Jeremy Stretch <jstretch@netboxlabs.com>

jeremystretch committed:
Changelog for #16424, #16523, #16654, #16657, #16689, #16714, #16723, #16725, #16735, #16747

github-actions[bot] committed:
Update source translation strings

jeremystretch committed:
Fixes #16679: Avoid overwriting custom JSON fields during bulk edit

Commit:
Closes #16716, add NAT IP to device view for OOB IP

jeremystretch committed:
Closes #16791: Add 200 & 400 Gbps selections for circuit termination port speed

jeremystretch committed:
Fixes #16807: Fix layout of VLAN edit form when custom fields are present

jeremystretch committed:
Fixes #16779: Fix saved filter selection for child object lists (#16789)

* Fixes #16779: Fix saved filter selection for child object lists

* Omit label_suffix

github-actions[bot] committed:
Update source translation strings

jeremystretch committed:
Fixes #16813: Fix ordering of bookmarks in dashboard widget when filtering by object type

jeremystretch committed:
Fixes #16806: Fix redirect URL when creating contact assignments with "add another" button

jeremystretch committed:
Fixes #16796: Allow assignment of VM with no site to a cluster with a site

jeremystretch committed:
Fixes #16780: IKE proposal created via REST API should not require authentication_algorithm

jeremystretch committed:
Changelog for #16679, #16716, #16779, #16780, #16791, #16796, #16806, #16807, #16813

AdamBrutsaert committed:
Closes #16817: Added 200 & 400 Gbps selections for circuit commit rate

jeremystretch committed:
Fixes #16721: Fix errant API request after deselecting a rack in device edit form

8ctorres committed:
Merge branch 'develop' of github.com:8ctorres/netbox into develop

8ctorres committed:
Fix automatically modified translations file

8ctorres committed:
Implemented corrections

8ctorres commented:
Hi @jeremystretch I rebased my branch onto the current develop branch, and reverted the changes to the "netbox/translations/en/LC_MESSAGES/django.po" (it's an automatic action, but still). Also implemented the changes you suggested.

I think it should be good now. Thanks for your patience.

github-actions[bot] committed:
Update source translation strings

jeremystretch commented:
@8ctorres it looks like something has gone wrong with this PR, as it includes a lot of changes that aren't in-scope for the issue. Please see if you can rebase your branch, or start a new one, and resubmit this PR.
